src/lib.rs

```diff
@@ -49,6 +49,50 @@ pub enum Error {
     InvalidId,
 }
 
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            #[cfg(feature = "remote_endpoint")]
+            Error::ChannelSend => write!(
+                f,
+                "event_manager: failed to send message to remote endpoint."
+            ),
+            #[cfg(feature = "remote_endpoint")]
+            Error::ChannelRecv => write!(
+                f,
+                "event_manager: failed to receive message from remote endpoint."
+            ),
+            #[cfg(feature = "remote_endpoint")]
+            Error::EventFd(_e) => write!(
+                f,
+                "event_manager: failed to manage EventFd file descriptor."
+            ),
+            Error::Epoll(_e) => write!(f, "event_manager: failed to manage epoll file descriptor."),
```

Comment:
Can we also print the inner error. It could be helpful for debugging things. Same comment for all the other errors.

Reply:
I like the way anyhow(https://crates.io/crates/anyhow) crate used to show error messages:)
We could get messages with different details by "{}", "{:?}" or "{:#?}".
Otherwise it always make me mad when dealing deeply nested errors.

Sample error message:
$ cargo run --quiet -- words.txt
Error: unable to count words in 'words.txt'

Caused by:
    0: Error encountered while reading from input
    1: read: broken pipe

Reply:
I am not sure I understand the proposal. Are you saying that debug already has that information and to hide it in Display?

Reply:
Recently I'm refactoring error handling related code and found that std::error::Error has two main member: description()/fmt() and cause()/source(). Per my understanding, fmt() should give information about the error it self, and cause()/source() are used to form an error chain.  I followed this design pattern.

On the other hand, the anyhow crate exports an `Error` structure, which are easy to use and provide convenient way to print the error message.
```
let e = anyhow::Error::new(low_level_error_with_chain);
format!("{}", e); // just show the description.
foramt!("{:?}", e) // print description and error chain backtrace.
format!("{:#?}", e) // print description and json-like error chain backtrace.
```

So I feel it may would be great if we:
1) remove all error!()/warn!() log messages from library crate
2) let the vmm decide how to the way to show the error message. And the anyhow crate is useful tool:)

Reply:
I think this makes sense. Just to be sure I understand, what you are proposing is some thing like this:
- The `Display` implementation for an error only generates a simple message, without recursively calling `Display` for the inner error (when present).
- The `Debug` derive (when present) works as before and prints the whole chain because it recursively uses `Debug` for every inner member of a type.
- VMMs/end users are free to use an external crate, or just create a function that uses the `Display` implementation and the `cause` method from the `Error` trait, to construct their own representation of the full error chain.

Reply:
Yes, this helps to generate nicer log messages.